File: 424.longest-repeating-character-replacement.java
// @leet start
class Solution {
    public int characterReplacement(String s, int k) {
        int charMap[] = new int[26];
        char strChar[] = s.toCharArray();
        int result = 0;
        int l = 0;
        char most = strChar[0];
        int replaceNum = 0;

        for (int r = 0; r < strChar.length; r++) {
            int currIdx = strChar[r] - 'A';
            charMap[currIdx]++;
            if (strChar[r] == most) {
                result = Math.max(result, r - l + 1);
                continue;
            }
            replaceNum++;
            while (replaceNum > k && l < r) {
                if (strChar[l] != most) {
                    replaceNum--;
                }

                int lIdx = strChar[l] - 'A';
                int count = --charMap[lIdx];
                l++;
            }

            int mostCount = charMap[most - 'A'];
            for (int i = 0; i < charMap.length; i++) {
                if (charMap[i] > mostCount) {
                    most = (char) (i + 'A');
                    mostCount = charMap[i];
                }
            }

            result = Math.max(result, r - l + 1);
        }

        return result;
    }
}
// @leet end

File: 424.longest-repeating-character-replacement.py
# @leet start
from collections import deque

class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        global_max = 0
        curr_count = 0
        curr_char = s[0]
        l = len(s)
        swaps = k
        char_list = deque()
        
        for i in range(1, l):
            if s[i] == curr_char:
                curr_count += 1
            elif swaps > 0:
                curr_count += 1
                swaps -= 1
            else:
                while len(char_list) > 0 and char_list[0] == curr_char:
                    curr_count -= 1
                    char_list.popleft()

                if len(char_list) > 0:
                    char_list.popleft()

            char_list.append(curr_char)
            global_max = max(global_max, curr_count)
        
        return global_max



# @leet end
